core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -546,36 +557,39 @@ V remove(K key, V oldValue, Function<ByteBuffer, V> transformer) {
             throw new NullPointerException();
         }
 
-        boolean logical = true; // when logical is false, means we have marked the handle as deleted
+        // when logicallyDeleted is true, it means we have marked the handle as deleted. Note that the entry may still be linked!
+        boolean logicallyDeleted = false;
         Handle prev = null;
         V v = null;
 
         while (true) {
-
             Chunk c = findChunk(key); // find chunk matching key
             Chunk.LookUp lookUp = c.lookUp(key);
-            if (lookUp != null && logical) {
-                prev = lookUp.handle; // remember previous handle
-            }
-            if (!logical && lookUp != null && prev != lookUp.handle) {
-                return v;  // someone else used this entry
-            }
 
             if (lookUp == null || lookUp.handle == null) {
-                return v; // there is no such key
+                // There is no such key. If we did logical deletion and someone else did the physical deletion,
+                // then the old value is saved in v. Otherwise v is (correctly) null
+                return v;
             }
 
-            if (logical) {
-                // we have marked this handle as deleted (successful remove)
+            if (logicallyDeleted) {
+                // This is the case where we logically deleted this entry (marked the handle as deleted), but someone reused
+                // the entry before we unlinked it. We have the previous value saved in v.
+                return v;
+            } else {
                 V vv = (transformer != null) ? (V) lookUp.handle.transform(transformer) : null;
```

Comment:
How do you know the value wouldn't be changed just after this line and before marked as deleted?

Reply:
It can be changed also after oldValue equals vv check...